Add `container-name-pattern`

nate10j opened the pull request:
<!-- Please answer the following. We close pull requests that don't. -->

> Which issue, if any, is this issue related to?

Closes #8432 

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

jeddy3 commented on lib/rules/container-name-pattern/README.md:
````diff
@@ -0,0 +1,48 @@
+
+<!-- prettier-ignore -->
+```css
+@container foo (max-width: 400px);
````
```suggestion
@container foo (width > 400px) {}
```

Let's use the more modern range context

jeddy3 commented on lib/rules/container-name-pattern/README.md:
````diff
@@ -0,0 +1,48 @@
+<!-- prettier-ignore -->
+```css
+@container foo (max-width: 400px);
+/**         ↑
````
```suggestion
/**        ↑
```
Alignment

jeddy3 commented on lib/rules/container-name-pattern/README.md:
````diff
@@ -0,0 +1,48 @@
+
+<!-- prettier-ignore -->
+```css
+@container bar {}
````
```suggestion
a { container-name: bar; }
```

Let's give examples of container names in declaration values.

jeddy3 commented on lib/rules/container-name-pattern/README.md:
````diff
@@ -0,0 +1,48 @@
+
+<!-- prettier-ignore -->
+```css
+@conainer FOO-bar {}
````
```suggestion
a { container: baz /  inline-size; }
```

jeddy3 commented on lib/rules/container-name-pattern/README.md:
````diff
@@ -0,0 +1,48 @@
+<!-- prettier-ignore -->
+```css
+@container foo-bar {}
+```
````
Let's add two patterns to match the new ones above:

```css
a { container-name: foo-bar; }
```

and 

```css
a { container: foo-baz /  inline-size; }
```

jeddy3 commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,63 @@
+		const regex = isString(primary) ? new RegExp(primary) : primary;
+
+		root.walkAtRules(atRuleRegexes.containerName, (containerNode) => {
+			if (!isStandardSyntaxAtRule(containerNode)) {
+				return;
+			}
```
```suggestion
			if (!isStandardSyntaxAtRule(containerNode)) return;
```

To align with our code style conventions.

jeddy3 commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,63 @@
+
+		const regex = isString(primary) ? new RegExp(primary) : primary;
+
+		root.walkAtRules(atRuleRegexes.containerName, (containerNode) => {
```
```suggestion
		root.walkAtRules(atRuleRegexes.containerName, (atRule) => {
```

To align with our code style conventions.

jeddy3 commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,63 @@
+				return;
+			}
+
+			const value = containerNode.params;
```
```suggestion
			const {params} = containerNode;
```

To align with our code style conventions of preferring destructuring.

jeddy3 commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,63 @@
+
+			const value = containerNode.params;
+
+			if (regex.test(value)) {
+				return;
+			}
```
```suggestion
			if (regex.test(value)) return;
```

To align with our code style conventions.

nate10j replied:
Hi thank you for reviewing.

jeddy3 reviewed (commented):
@nate10j Thank you for contributing and starting on this. It's looking good so far.

The rule will need to:

- check container names within declaration values
- account for complete container conditions

Container names can appear in the [`container-name`](https://drafts.csswg.org/css-conditional-5/#container-name) property and the [`container`](https://drafts.csswg.org/css-conditional-5/#container-shorthand) shorthand property. You can use the value parser to find the container name within the value of these properties.

Container conditions have a [broad syntax](https://drafts.csswg.org/css-conditional-5/#container-rule).

```
@container <container-condition># {
  <block-contents>
}
```

And container names can appear zero or multiple times within a `@container` prelude, e.g. `foo` and `bar` in:

```css
@container foo (inline-size > 30em), bar style(--large: true) {}
```

I think we can use the value parser and keep the logic naive by assuming that any top-level `word` node that isn't `none`, `or`, `and`, or `not` is a container name.

You can use [`custom-property-pattern`](https://github.com/stylelint/stylelint/blob/main/lib/rules/custom-property-pattern/index.mjs) as a reference for:

- checking both declarations and at-rules
- using the value parser

I've also added some commitable suggestions as part of the first review.

Thanks again for contributing! It's a rule that many people will benefit from as we'll turn it on in our standard config.

Mouvedia commented on lib/rules/container-name-pattern/__tests__/index.mjs:
Could you add test cases that use `style()` and `scroll-state()`?
e.g. https://drafts.csswg.org/css-conditional-5/#example-cba8e10c

Could you also add a passing `@container () {}`?
i.e. nameless

nate10j replied:
I have made your required changes. Let me know if you want me to add more testcases.

jeddy3 replied:
@nate10j Thanks for adding more test cases.

Can you also add ones for a:

- condition without a name as names are optional: `@container (width > 400px) {}`
- container name in the value of a [`container-name`](https://drafts.csswg.org/css-conditional-5/#container-name) property: `a { container-name: bar; }`
- container name in the value of a [`container`](https://drafts.csswg.org/css-conditional-5/#container-name) shorthand property: `a { container: baz / inline-size; }`

Ref: https://github.com/stylelint/stylelint/pull/8498#pullrequestreview-2727869222

nate10j replied:
I have added support for nameless conditions and multiple names in my recent commits.

nate10j replied:
Hi, I have a question. How can I use ```walkDecls``` for more than one property? (property and shorthand property)? I tried looking through ``rejexes.mjs`` to see find any other references.

jeddy3 replied:
@nate10j You can use a regex. For example, this one checks the `font-weight` property and the `font` shorthand:

https://github.com/stylelint/stylelint/blob/27439d936707b660591334fc1fe01857474d3056/lib/rules/font-weight-notation/index.mjs#L67

> I tried looking through rejexes.mjs to see find any other references.

We only recently started moving regexes in the `regexes.mjs` file. Let's add ours there, e.g:

```js
export const propertyRegexes = {
	containerNameAndShorthandName: /^container(-name)?$/i,
};
```

jeddy3 commented on lib/rules/container-name-pattern/__tests__/index.mjs:
```diff
@@ -0,0 +1,111 @@
+			endLine: 1,
+			endColumn: 19,
+		},
+		{
+			code: '@container foo-bar (), FOO-bar () {}',
+			message: messages.expected('FOO-bar', 'foo-.+'),
+			line: 1,
+			column: 24,
+			endLine: 1,
+			endColumn: 31,
+		},
```
Let's add a testcase with two warnings:

```css
@container bar (inline-size > 30em), baz style(--large: true) {}
```

[example](https://github.com/stylelint/stylelint/blob/cf4c7acc616a0366d17891ea93877c62659d12d3/lib/rules/alpha-value-notation/__tests__/index.mjs#L198-L213)

nate10j replied:
Hi. I have added a testcase for two warnings.

nate10j commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,104 @@
+
+				if (node.type !== 'word') return;
+
+				if (!isStandardSyntaxContainerName(node.value)) return;
```
Hi. Thank you for reviewing. I have deleted isStandardSyntax and used a this implementation. I have also made the changes above and used a regex to denote container names and their shorthand names, deleted unnecessary tests.

I have added required tests of nameless containers, container declaration for both shorthand and container name.

Let me know if you need me to make any more changes.

jeddy3 commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,105 @@
+
+			const parsedValue = valueParser(params);
+
+			let functions = [];
+
+			parsedValue.walk((node) => {
+				if (node.type === 'function') functions.push(node);
+
+				if (node.type !== 'word') return false;
+
+				if (NON_CONTAINER_NAMES.has(node.value.toLowerCase())) return;
+
+				for (const child of functions) {
+					if (child.nodes.includes(node)) return;
+				}
+
+				if (regex.test(node.value)) return;
+
+				const index = atRuleParamIndex(atRule) + node.sourceIndex;
+
+				complain(index, node.value, atRule);
+			});
```
```suggestion
			parsedValue.walk((node) => {
				if (node.type !== 'word') return false;

				if (KEYWORDS.has(node.value.toLowerCase())) return;

				if (regex.test(node.value)) return;

				const index = atRuleParamIndex(atRule) + node.sourceIndex;

				complain(index, node.value, atRule);
			});
```

Let's refactor as stopping the traverse into descendant nodes using `return false` makes the function code redundant.

jeddy3 commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,105 @@
+	url: 'https://stylelint.io/user-guide/rules/container-name-pattern',
+};
+
+const NON_CONTAINER_NAMES = new Set(['and', 'or', 'none', 'not']);
```
```suggestion
const KEYWORDS = new Set(['and', 'or', 'none', 'not']);
```

Let's rename this var as these are keywords, whereas the container names are custom-idents.

See [specs](https://drafts.csswg.org/css-conditional-5/#container-name).

jeddy3 commented on lib/rules/container-name-pattern/__tests__/index.mjs:
```diff
@@ -0,0 +1,133 @@
+			code: '@container foo-bar {}',
+			description: 'Accepts pattern in RegExp notation',
+		},
+		{
+			code: '@container () {}',
+			description: 'nameless container condition',
+		},
```
```suggestion
```
Let's remove this redundant test.

jeddy3 commented on lib/rules/container-name-pattern/__tests__/index.mjs:
```diff
@@ -0,0 +1,133 @@
+			`,
+		},
+		{
+			code: stripIndent`
+				a { container: foo-bar / inline-size; }
+			`,
```
```suggestion
			code: 'a { container: foo-bar / inline-size; }',
```

jeddy3 commented on lib/rules/container-name-pattern/__tests__/index.mjs:
```diff
@@ -0,0 +1,133 @@
+		},
+
+		{
+			code: stripIndent`
+				.my-component {
+					container-name: foo-bar;
+				}
+			`,
```
```suggestion
			code: 'a { container-name: foo-bar; }'
```

jeddy3 replied:
Let's update the tests to match the [conventions](https://stylelint.io/developer-guide/rules#write-tests).

jeddy3 commented on lib/rules/container-name-pattern/__tests__/index.mjs:
```diff
@@ -0,0 +1,133 @@
+			endColumn: 31,
+		},
+		{
+			code: stripIndent`
+				.my-component {
+					container-name: FOO-bar;
+				}
+			`,
```
```suggestion
			code: 'a { container-name: FOO-bar; }',
```

jeddy3 commented on lib/rules/container-name-pattern/__tests__/index.mjs:
```diff
@@ -0,0 +1,133 @@
+	config: ['foo-.+'],
+
+	accept: [
+		{ code: '@container foo-bar {}' },
```
```suggestion
		{ code: 'a { color: red; }' },
		{ code: '@container NONE {}' },
		{ code: '@container foo-bar {}' },
```

Let's add tests for the:

- decl regex
- `none` keyword

jeddy3 commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,105 @@
+
+		const regex = isString(primary) ? new RegExp(primary) : primary;
+
+		root.walkDecls(propertyRegexes.containerNameAndShorthandName, (decl) => {
```
```suggestion
		root.walkDecls(propertyRegexes.containerNameAndShorthandName, (decl) => {
		  if (!isStandardSyntaxDecl(decl)) return;
```

Let's only check standard declarations.

jeddy3 commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,105 @@
+
+			if (!firstNode) return;
+
+			if (!isStandardSyntaxProperty(firstNode.value)) return;
```
```suggestion
			if (!isStandardSyntaxValue(firstNode.value)) return;
```

Let's change this as `firstNode.value` is a value, not a property.

jeddy3 commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,105 @@
+
+import { atRuleRegexes, propertyRegexes } from '../../utils/regexes.mjs';
+import { isRegExp, isString } from '../../utils/validateTypes.mjs';
+import { atRuleParamIndex } from '../../utils/nodeFieldIndices.mjs';
+import { declarationValueIndex } from '../../utils/nodeFieldIndices.mjs';
```
```suggestion
import { atRuleParamIndex, declarationValueIndex } from '../../utils/nodeFieldIndices.mjs';
```

jeddy3 reviewed (changes requested):
@nate10j Thank you for making the changes. It's shaping up nicely.

I've requested some more refactors and tests based on the last round of changes.

nate10j commented:
Hi. I have made further changes, and additionally added missing import statements, and tweaked expected results after removing ```stripIndent```. All test cases work now.

jeddy3 commented on lib/rules/container-name-pattern/__tests__/index.mjs:
```diff
@@ -0,0 +1,120 @@
+			code: '@container () {}',
+			description: 'nameless container condition',
+		},
+
```
```suggestion
```

jeddy3 commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,97 @@
+			possible: [isRegExp, isString],
+		});
+
+		if (!validOptions) {
+			return;
+		}
```
```suggestion
		if (!validOptions) return;
```

jeddy3 reviewed (approved):
@nate10j Excellent, thank you. LGTM.

(We typically wait for a 2nd review on new rules before merging, as a second pair of eyes usually spots things the first missed.)

Mouvedia commented on lib/rules/container-name-pattern/__tests__/index.mjs:
```diff
@@ -0,0 +1,119 @@
+	accept: [
+		{
+			code: '@container foo-bar {}',
+			description: 'Accepts pattern in RegExp notation',
+		},
+	],
+
+	reject: [
+		{
+			code: '@container foo-baz {}',
+			description: 'Accepts pattern in RegExp notation',
```
question

Both the reject and the accept share the same description, is that on purpose?

jeddy3 replied:
Let's remove the 2nd one.

Mouvedia commented on lib/rules/container-name-pattern/__tests__/index.mjs:
Can you add a test for multiple names?
e.g. `container-name: foo bar;`

jeddy3 replied:
Good catch! I missed the `+` in the [specs](https://drafts.csswg.org/css-conditional-5/#container-name).

Let's also add tests for multiple names in the shorthand:

```css
a { container: foo bar / inline-size; }
```

This will involve updating the logic to check all the `word` nodes before `/` rather than just the first node.

Mouvedia replied:
> And let's add tests for global keywords…

👍

ybiquitous reviewed (approved):
Overall LGTM 👍🏼 

I'm okay with merging this after addressing the rest reviews by @Mouvedia and @jeddy3.

nate10j commented:
Hi. Thanks for reviewing. I added support for multiple names in shorthand and global keywords, and added required tests for them.

For global keywords, I added another message
```javascript
const messages = ruleMessages(ruleName, {
	expected: (containerName, pattern) => `Expected "${containerName}" to match pattern "${pattern}"`,
	rejected: (globalKeyword) => `"${globalKeyword}" is a global keyword`,
});
```
I am wondering if the global keywords should be an error message or just be simply ignored like the example below:
```javascript
if (basicKeywords.has(value.toLowerCase())) return;
```

ybiquitous commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,143 @@
+					});
+
+					return;
+				}
```
[suggestions] As Codecov says, `if (basicKeywords.has(value.toLowerCase())) { ... }` block doesn't seem to be tested. Can you add a test case to improve the coverage? 🙏🏼

nate10j replied:
Hi. I have added a testcase for shorthand container for ``if (basicKeywords.has(value.toLowerCase()))`` and it should improve the coverage.

jeddy3 commented:
> I am wondering if the global keywords should be an error message or just be simply ignored like the example below:

They should simply be ignored as they are valid values (that aren't container names).

jeddy3 commented on lib/rules/container-name-pattern/__tests__/index.mjs:
```diff
@@ -0,0 +1,168 @@
+	config: ['foo-.+'],
+
+	accept: [
+		{ code: 'a { color: red; }' },
```
```suggestion
		{ code: 'a { color: red; }' },
		a { container-name: initial; }
```

jeddy3 commented on lib/rules/container-name-pattern/__tests__/index.mjs:
```diff
@@ -0,0 +1,168 @@
+			endLine: 1,
+			endColumn: 28,
+		},
+		{
+			code: 'a { container-name: initial; }',
+			message: messages.rejected('initial'),
+			line: 1,
+			column: 21,
+			endLine: 1,
+			endColumn: 28,
+		},
```
```suggestion
```

nate10j commented:
Hi. Thanks for reviewing. I have made changes. I removed the error messages and instead ignored it with a return statement.

jeddy3 commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,118 @@
+
+		root.walkDecls(propertyRegexes.containerName, (decl) => {
+			if (!isStandardSyntaxDeclaration(decl)) return;
+
+			const parsedValue = valueParser(decl.value);
+
+			parsedValue.walk(({ sourceIndex, type, value }) => {
+				if (type !== 'word') return false;
+
+				if (basicKeywords.has(value.toLowerCase())) return;
+
+				if (!isStandardSyntaxValue(value)) return;
+
+				if (regex.test(value)) return;
+
+				complain(declarationValueIndex(decl) + sourceIndex, value, decl);
+			});
+		});
+
+		root.walkDecls(propertyRegexes.containerNameShorthand, (decl) => {
+			if (!isStandardSyntaxDeclaration(decl)) return;
+
+			const parsedValue = valueParser(decl.value);
+
+			parsedValue.walk((node) => {
+				let { sourceIndex, type, value } = node;
+
+				if (type !== 'word') return false;
+
+				if (node === parsedValue.nodes[parsedValue.nodes.length - 1]) return;
+
+				if (basicKeywords.has(value.toLowerCase())) return;
+
+				if (!isStandardSyntaxValue(value)) return;
+
+				if (regex.test(value)) return;
+
+				complain(declarationValueIndex(decl) + sourceIndex, value, decl);
+			});
+		});
```
Let's recombine these into a single block that checks the `container` and `container-name` properties to remove the repetition and address the multiple container types test.

> This will involve updating the logic to check all the `word` nodes before `/` rather than just the first node.

There may be a more eloquent way of doing this, but this should work:

```js
let isContainerType = false;

parsedValue.walk((node) => {
	let { sourceIndex, type, value } = node;

	if (isContainerType) return;

	if (type === 'div' && value === '/') isContainerType = true;

	if (type !== 'word') return false;
```

jeddy3 reviewed (changes requested):
Thank you for making the changes. I've suggested another edge case that we'll need to consider.

nate10j commented:
Hi. Thanks for reviewing. I have made required changes.

jeddy3 commented on lib/rules/container-name-pattern/index.mjs:
```diff
@@ -0,0 +1,102 @@
+
+		if (!validOptions) return;
+
+		const regex = isString(primary) ? new RegExp(primary) : primary;
```
```suggestion
		const regex = isString(primary) ? new RegExp(primary) : primary;
		
		const languageCssWideKeywords =
			result.stylelint.config?.languageOptions?.syntax?.cssWideKeywords ?? [];

		languageCssWideKeywords.forEach(basicKeywords.add, basicKeywords);
```

jeddy3 commented on lib/rules/container-name-pattern/__tests__/index.mjs:
```diff
@@ -0,0 +1,167 @@
+			endColumn: 19,
+		},
+	],
+});
```
```suggestion
});

testRule({
	ruleName,
	config: ['foo-.+'],
	languageOptions: {
		syntax: {
			cssWideKeywords: ['bar'],
		},
	},

	accept: [
		{
			code: '@container foo-bar {}',
		},
		{
			code: 'a { container-name: bar; }',
		},
	],

	reject: [
		{
			code: 'a { container-name: foo; }',
			message: messages.expected('foo', 'foo-.+'),
			line: 1,
			column: 21,
			endLine: 1,
			endColumn: 24,
		},
	],
});
```

nate10j replied:
Hi. I have committed the suggestions. Do you need the ``languageOptions`` to be in the first ``testRule`` also? Let me know if there needs any further changes.

jeddy3 replied:
> Do you need the `languageOptions` to be in the first testRule also?

No, thank you. It's intentionally in a separate `testRule` so we can test with _and without_ `languageOptions` being set.

jeddy3 reviewed (changes requested):
It's looking great, thank you.

One last thing, we'll need to account for [`languageOptions`](https://stylelint.io/user-guide/configure/#languageoptions) because we're now ignoring global keywords.

I've suggested some changes and additional tests for this.

We'll need to run:

```console
npm i -D jest-preset-stylelint@latest
```

As part of this PR because we only recently added support for testing `languageOptions` to our jest preset.

As an aside, thank you for your patience dealing with these changes. It's typical that there are a few rounds of changes when implementing a new rule, as we uncover and address edge cases.

jeddy3 reviewed (approved):
LGTM, thank you.

(We can patch up the unrelated failing tests in another PR.)

ybiquitous commented:
@jeddy3 The update of `jest-preset-stylelint` fails CI. See #8510 👀 
(I manually triggered Dependabot)

ybiquitous commented:
#8510 updated `jest-preset-stylelint`, fixed the failed test, and was merged.
I've merged the main branch into this one. Let's see the CI results!

ybiquitous reviewed (approved):
CI is now recovered ✅
